[JENKINS-33800] - fileNotFound exception / more accurately determine if this is an upgrade

kzantow opened the pull request:
If the Jenkins 1 global settings are never saved and no jobs were created, the setup wizard may treat this case as a new installation. In the case the security config is saved, this causes an issue - this more "accurately" determines if this instance is actually an upgrade vs. a new install and prevents a fileNotFound exception looking for the setup token.

This addresses issue: https://issues.jenkins-ci.org/browse/JENKINS-33800

@reviewbybees

kzantow committed:
JENKINS-33800 - fileNotFound exception if jenkins 1 installed, security
settings saved but global config is not - show normal login in this case

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

kzantow commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -107,23 +107,25 @@ public SetupWizard(Jenkins j) throws IOException, InterruptedException {
-                + "*************************************************************" + ls
-                + "*************************************************************" + ls);
+        if(iapf.exists()) {
+            String setupKey = iapf.readToString().trim();
```
Only whitespace here - indented

daniel-beck commented:
In the PR build, this results in the setup wizard being entered without needing authentication, and then failing to do anything as the user is missing Overall/Administer. I think the optimal way here would be more comprehensive update detection, so the setup wizard isn't even invoked unless it's a pristine instance.

kzantow committed:
Issue with timing determining install state; mark jenkins as upgrade if
jobs exist or security configuration was set up

jglick commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -24,10 +24,11 @@ THE SOFTWARE.
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <j:if test="${it.setupWizard != null}">
+  <j:choose>
+  <j:when test="${it.setupWizard.useGeneratedPassword()}">
```
So `it.setupWizard` is now unconditionally nonnull? I did not see that change made on the Java side.

jglick replied:
And I still see a null check in `loginError.jelly` so probably :bug:

kzantow replied:
@jglick it definitely may be null... normal EL behavior is `it.setupWizard` being null short-circuits the expression to false, though... (and it seems to be doing that here, am I missing something)

kzantow committed:
Merge remote-tracking branch 'primary/2.0' into JENKINS-33800-initial-password-file-not-found

kzantow committed:
Update loginError.jelly to match login.jelly setup wizard check

daniel-beck commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -88,7 +89,8 @@ public static InstallState getInstallState() {
             // they actually did anything
-            if (!Jenkins.getInstance().getItemMap().isEmpty()) {
+            if (!Jenkins.getInstance().getItemMap().isEmpty()
+                    || Jenkins.getInstance().getSecurityRealm() != SecurityRealm.NO_AUTHENTICATION) {
```
Does this correctly handle the 2.0 case, which is a different security realm by default?

kzantow replied:
@daniel-beck it handles the normal 2.0 case but restarts may be an issue, I'll investigate this

kzantow committed:
Merge remote-tracking branch 'primary/2.0' into JENKINS-33800-initial-password-file-not-found

kzantow committed:
Fix setup wizard cases: update wizard may show erroneously,
crash/restart during install may result in installer not completed,
ensure the user may not get locked out if they've created an admin user
but not completed the installer

jglick commented:
:bee: without understanding all of it

jtnord commented:
:bee:

kzantow commented:
Hmm looks like failing tests might be related, investigating.

kzantow committed:
Merge remote-tracking branch 'primary/2.0' into JENKINS-33800-initial-password-file-not-found

kzantow committed:
Fix tests, move forced update site download to setup wizard

kzantow commented:
@jtnord @jglick @daniel-beck  please re-review, there are some changes to address test failures

kzantow commented:
Looks like tests passed, what the heck?

Tests run: 634, Failures: 0, Errors: 0, Skipped: 9

jtnord commented:
@kzantow surefire process was killed/died.  So not all of the tests have run :)

```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project test: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?```

You should see something like
```Tests run: 8330, Failures: 0, Errors: 0, Skipped: 11```  which is a whopping 7000+ more tests :-)

kzantow commented:
... ah right. D'oh.

jglick commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -126,6 +129,13 @@ public SetupWizard(Jenkins j) throws IOException, InterruptedException {
+        try {
+            // Make sure plugin metadata is up to date
+            UpdateCenter.updateDefaultSite();
+        } catch (Exception e) {
```
:ant: Huh? That method is already `catch`ing `Exception` and logging it.

kzantow replied:
Hm, I suppose so... mainly I don't want it to fail here, it will try again when the wizard is accessed if it hasn't had a successful site update. I might rather leave it as-is in case someone decides to change the `updateDefaultSite()` without considering it being used here.

jglick commented:
:bee: without real comprehension

jtnord commented:
re:bee:

kzantow commented:
@reviewbybees done thanks!

amuniz commented:
Tested by upgrading from 1.624.4. It works fine.
:bee: Great job all around the install wizard @kzantow!!

amuniz commented:
Tested upgrading from 1.656 (last weekly). Works fine also.

kzantow commented:
@daniel-beck is this sufficiently reviewed for your liking to merge?
